### ISSUE TITLE
Installing extension modules in setup.py?

### ISSUE THREAD

**Issue by croth1** [CGATOxford/UMI-tools#126]
Is there a reason for not building the extension modules in the `setup.py` file, but generate them on the fly when needed?

This seems to give me permission errors when working with a python interpreter that I do not have write access to:

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.5/site-packages/pyximport/pyximport.py", line 216, in load_module
    inplace=build_inplace, language_level=language_level)
  File "/opt/conda/lib/python3.5/site-packages/pyximport/pyximport.py", line 192, in build_module
    reload_support=pyxargs.reload_support)
  File "/opt/conda/lib/python3.5/site-packages/pyximport/pyxbuild.py", line 102, in pyx_to_dll
    dist.run_commands()
  File "/opt/conda/lib/python3.5/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/opt/conda/lib/python3.5/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/opt/conda/lib/python3.5/site-packages/Cython/Distutils/old_build_ext.py", line 185, in run
    _build_ext.build_ext.run(self)
  File "/opt/conda/lib/python3.5/distutils/command/build_ext.py", line 338, in run
    self.build_extensions()
  File "/opt/conda/lib/python3.5/site-packages/Cython/Distutils/old_build_ext.py", line 192, in build_extensions
    ext.sources = self.cython_sources(ext.sources, ext)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Distutils/old_build_ext.py", line 345, in cython_sources
    full_module_name=module_name)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Main.py", line 684, in compile
    return compile_single(source, options, full_module_name)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Main.py", line 634, in compile_single
    return run_pipeline(source, options, full_module_name)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Main.py", line 494, in run_pipeline
    err, enddata = Pipeline.run_pipeline(pipeline, source)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Pipeline.py", line 340, in run_pipeline
    data = phase(data)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Pipeline.py", line 53, in generate_pyx_code_stage
    module_node.process_implementation(options, result)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/ModuleNode.py", line 137, in process_implementation
    self.generate_c_code(env, options, result)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/ModuleNode.py", line 387, in generate_c_code
    f = open_new_file(result.c_file)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Utils.py", line 65, in open_new_file
    return codecs.open(path, "w", encoding="ISO-8859-1")
  File "/opt/conda/lib/python3.5/codecs.py", line 895, in open
    file = builtins.open(filename, mode, buffering)
PermissionError: [Errno 13] Permission denied: '/opt/conda/lib/python3.5/site-packages/umi_tools/_dedup_umi.c'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.5/site-packages/umi_tools/network.py", line 20, in <module>
    from umi_tools._dedup_umi import edit_distance
  File "/opt/conda/lib/python3.5/site-packages/pyximport/pyximport.py", line 445, in load_module
    language_level=self.language_level)
  File "/opt/conda/lib/python3.5/site-packages/pyximport/pyximport.py", line 232, in load_module
    raise exc.with_traceback(tb)
  File "/opt/conda/lib/python3.5/site-packages/pyximport/pyximport.py", line 216, in load_module
    inplace=build_inplace, language_level=language_level)
  File "/opt/conda/lib/python3.5/site-packages/pyximport/pyximport.py", line 192, in build_module
    reload_support=pyxargs.reload_support)
  File "/opt/conda/lib/python3.5/site-packages/pyximport/pyxbuild.py", line 102, in pyx_to_dll
    dist.run_commands()
  File "/opt/conda/lib/python3.5/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/opt/conda/lib/python3.5/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/opt/conda/lib/python3.5/site-packages/Cython/Distutils/old_build_ext.py", line 185, in run
    _build_ext.build_ext.run(self)
  File "/opt/conda/lib/python3.5/distutils/command/build_ext.py", line 338, in run
    self.build_extensions()
  File "/opt/conda/lib/python3.5/site-packages/Cython/Distutils/old_build_ext.py", line 192, in build_extensions
    ext.sources = self.cython_sources(ext.sources, ext)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Distutils/old_build_ext.py", line 345, in cython_sources
    full_module_name=module_name)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Main.py", line 684, in compile
    return compile_single(source, options, full_module_name)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Main.py", line 634, in compile_single
    return run_pipeline(source, options, full_module_name)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Main.py", line 494, in run_pipeline
    err, enddata = Pipeline.run_pipeline(pipeline, source)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Pipeline.py", line 340, in run_pipeline
    data = phase(data)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/Pipeline.py", line 53, in generate_pyx_code_stage
    module_node.process_implementation(options, result)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/ModuleNode.py", line 137, in process_implementation
    self.generate_c_code(env, options, result)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Compiler/ModuleNode.py", line 387, in generate_c_code
    f = open_new_file(result.c_file)
  File "/opt/conda/lib/python3.5/site-packages/Cython/Utils.py", line 65, in open_new_file
    return codecs.open(path, "w", encoding="ISO-8859-1")
  File "/opt/conda/lib/python3.5/codecs.py", line 895, in open
    file = builtins.open(filename, mode, buffering)
ImportError: Building module umi_tools._dedup_umi failed: ["PermissionError: [Errno 13] Permission denied: '/opt/conda/lib/python3.5/site-packages/umi_tools/_dedup_umi.c'\n"]
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.5/site-packages/umi_tools/dedup.py", line 279, in <module>
    import umi_tools.network as network
  File "/opt/conda/lib/python3.5/site-packages/umi_tools/network.py", line 24, in <module>
    from _dedup_umi import edit_distance
ImportError: No module named '_dedup_umi'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/conda/bin/umi_tools", line 11, in <module>
    load_entry_point('umi-tools==0.4.3', 'console_scripts', 'umi_tools')()
  File "/opt/conda/lib/python3.5/site-packages/umi_tools/umi_tools.py", line 43, in main
    module = imp.load_module(command, file, pathname, description)
  File "/opt/conda/lib/python3.5/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/opt/conda/lib/python3.5/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 693, in _load
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 665, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/opt/conda/lib/python3.5/site-packages/umi_tools/dedup.py", line 281, in <module>
    import network
ImportError: No module named 'network'
```

**Comment by IanSudbery** [comment]
Currently if you needed root to install, you'll need root the first time you run (you can just run --help). The reason we currently run things this way is to avoid dependency order problems (i.e. to avoid have install dependencies, rather than just run dependencies). We may revisit this.

**Comment by croth1** [comment]
Excuse my followup question here, unfortunately I'm not familiar with the build system involved here:

`umi_tools dedup --help` creates `_dedup_umi.c` for me, but does not create a shared library yet. Will that work, or will it need a second write access for compiling the a `.so`  file?

**Comment by andreas-wilm** [comment]
Hi all, just wanted to bump this thread. I ran into a similar problem (ImportError: No module named network) when running dedup as user for the first time (installation was done as admin). This was solved by running dedup once as admin. This is obviously confusing for users and also an awkward solution. But it does work :)

Andreas

**Comment by IanSudbery** [comment]
We will try and get this done in 0.5

**Comment by IanSudbery** [comment]
Dear @croth1 and @andreas-wilm, we think that these issues have been solved in 0.5. 

* Extensions are now compiled at install time
* Errors should now not show up as import errors of unrelated modules.

We are closing this for now, please feel free to reopen if you are still having issues.

**Comment by andreas-wilm** [comment]
Thanks
